vm-validator/src/vm_validator.rs

```diff
@@ -121,3 +122,47 @@ pub fn get_account_sequence_number(
         None => Ok(0),
     }
 }
+
+// A pool of VMValidators that can be used to validate transactions concurrently. This is done because
+// the VM is not thread safe today. This is a temporary solution until the VM is made thread safe.
+#[derive(Clone)]
+pub struct PooledVMValidator {
+    vm_validators: Vec<Arc<Mutex<VMValidator>>>,
+}
+
+impl PooledVMValidator {
+    pub fn new(db_reader: Arc<dyn DbReader>, pool_size: usize) -> Self {
+        let mut vm_validators = Vec::new();
+        for _ in 0..pool_size {
+            vm_validators.push(Arc::new(Mutex::new(VMValidator::new(db_reader.clone()))));
+        }
+        PooledVMValidator { vm_validators }
+    }
+
+    pub fn get_next_vm(&self) -> Arc<Mutex<VMValidator>> {
+        let mut rng = thread_rng(); // Create a thread-local random number generator
+        let random_index = rng.gen_range(0, self.vm_validators.len()); // Generate random index
+        self.vm_validators[random_index].clone() // Return the VM at the random index
```

Comment:
What's your considerations behind this design (vs thread locals, for example)?

Reply:
IMO, this is a cleaner design as it implements a generic pool, which also allows flexibility of initializing a specific number of VMs irrespective of number of threads.

Reply:
well, sounds like unnecessary blocking to me, but yeah, probably only "potential" blocking  :D

Reply:
Silly question: instead of using random numbers, why not just round robin the validators? (e.g., maintain an atomic u64 to point to the next VM index to use and wrap around?). This might reduce unnecessary blocking if we aren't using more VMValidators than threads at any given time?

Reply:
I had done the same thing in the previous commit (only with global counter protected by Mutex) - https://github.com/aptos-labs/aptos-core/pull/13091/commits/31082bebd2013a25a6b57d722424d0119e603d1a. I realized it adds unnecessary blocking. Even I tried with atomic u64 but atomically increasing the pointer and wrapping around makes it ugly and not worth it.